File: OOPCoursework/SpaceWar/OOPCoursework/Planet.cpp
#include "stdafx.h"
#include "Planet.h"
#include "Regenship.h"
#include "Stealthship.h"
#include "Ship_radar.h"
#include "Transportship.h"

Planet::Planet(void){
	Resource = 0;
	Captured = false;
	Visibility = 0;
	icon = Image::FromFile("PlanetPic.png");
}

Planet::Planet(Point^ _location,Graph^ _graph,const Int32 _res,const Boolean _occup, const Int32 _visibility): Rootclass(_location,_graph){
	Resource = _res;
	Captured = _occup;
	Visibility = _visibility;
	icon = Image::FromFile("PlanetPic.png");
}

Planet::Planet(Planet% copyobj): Rootclass(copyobj){
	Resource = copyobj.Resource;
	Captured = copyobj.Captured;
	Visibility = copyobj.Visibility;
	icon = Image::FromFile("PlanetPic.png");
}

Void Planet::Resource::set(Int32 Value){
	resource = Value;
}

Void Planet::Visibility::set(Int32 Value){
	visibility = Value;
}

Void Planet::Captured::set(Boolean Value){
	captured = Value;
}

Int32 Planet::Resource::get(){
	return resource;
}

Int32 Planet::Visibility::get(){
	return visibility;
}

Boolean Planet::Captured::get(){
	return captured;
}

String^ Planet::ToString(){
	String^ string = gcnew String("");
	string += "Planet: R "; string += Resource.ToString();
	string += " Occ "; string += Captured.ToString();
	string += " L("; string += Location->X.ToString();
	string += ";"; string += Location->Y.ToString();
	string += ")";
	return string;
}

Rootclass^ Planet::create_warship(){
	if(Resource >= 150){
		Resource -= 150;
		return gcnew Warship(Location,graph,3,250,20,5);
	}
	else return nullptr;
}

Rootclass^ Planet::create_ship_radar(){
	if(Resource >= 100){
		Resource -= 100;
		return gcnew Ship_radar(Location,graph,5,100,25);
	}
	else return nullptr;
}

Rootclass^ Planet::create_transportship(){
	if(Resource >= 75){
		Resource -= 75;
		return gcnew Transportship(Location,graph,6,80,20);
	}
	else return nullptr;
}

Rootclass^ Planet::create_regenship(){
	if(Resource >= 200){
		Resource -= 200;
		return gcnew Regenship(Location,graph,3,200,20,6);
	}
	else return nullptr;
}

Rootclass^ Planet::create_stealthship(){
	if(Resource >= 180){
		Resource -= 180;
		return gcnew Stealthship(Location,graph,4,150,10,8);
	}
	else return nullptr;
}

Void Planet::draw(Graphics^ canvas,Rectangle^ canvassize, Rectangle^ cursor){
	RectangleF loc;
	loc.X = canvassize->Width*((Location->X - cursor->X)/Double(cursor->Width));
	loc.Y = canvassize->Height*((Location->Y - cursor->Y)/Double(cursor->Height));
	loc.Width = canvassize->Width/Double(cursor->Width);
	loc.Height = canvassize->Height/Double(cursor->Height);
	canvas->DrawImage(icon,loc);
}

File: Practices/SpaceWar/SpaceWar/Planet.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Threading;
using System.Drawing;

namespace SpaceWar
{
    class Planet: LocatableObject, DisableStealth
    {
        #region Attributes
        int resources;
        int visibility;
        PlayerColor playerColor;
        Stack<PlayerColor> capture;
        PlanetAnimationImages animations;
        Space space;
        #endregion

        public int Vision()
        {
            return visibility;
        }

        public Planet(int x, int y, Space _space, string[][] parametrs, PlayerColor color = PlayerColor.None): base(x,y)
        {
            playerColor = color;
            space = _space;
            for (int i = 0; i < parametrs.Length; i++)
            {
                switch (parametrs[i][0])
                {
                    case "Visibility": visibility = Convert.ToInt32(parametrs[i][1]); break;
                    case "Resources": Random rand = new Random();
                        resources = rand.Next(Convert.ToInt32(parametrs[i][1]), Convert.ToInt32(parametrs[i][2])); break;
                }
            }
        }

        public PlayerColor Color
        {
            get 
            {
                return playerColor;
            }
        }

        public int Resources
        {
            get
            {
                return resources;
            }
        }

        public LocatableObject CreateWarShip(int price, string[][] parameters)
        {
            resources -= price;
            int x = Position.X + 1, y = Position.Y;
            if (space[Position.X + 1, Position.Y] != null)
                (space[Position.X + 1, Position.Y] as Tactic).Retreat(this);
            space[Position.X + 1, Position.Y] = new WarShip(x, y, playerColor, space, parameters);
            return space[Position.X + 1, Position.Y];
        }

        public LocatableObject CreateStealthShip(int price, string[][] parameters)
        {
            resources -= price;
            int x = Position.X + 1, y = Position.Y;
            if (space[Position.X + 1, Position.Y] != null)
                (space[Position.X + 1, Position.Y] as Tactic).Retreat(this);
            space[Position.X + 1, Position.Y] = new StealthShip(x, y, playerColor, space, parameters);
            return space[Position.X + 1, Position.Y];
        }

        public LocatableObject CreateRegenShip(int price, string[][] parameters)
        {
            resources -= price;
            int x = Position.X + 1, y = Position.Y;
            if (space[Position.X + 1, Position.Y] != null)
                (space[Position.X + 1, Position.Y] as Tactic).Retreat(this);
            space[Position.X + 1, Position.Y] = new RegenShip(x, y, playerColor, space, parameters);
            return space[Position.X + 1, Position.Y];
        }

        public LocatableObject CreateTransportShip(int price, string[][] parameters)
        {
            resources -= price;
            int x = Position.X + 1, y = Position.Y;
            if (space[Position.X + 1, Position.Y] != null)
                (space[Position.X + 1, Position.Y] as Tactic).Retreat(this);
            space[Position.X + 1, Position.Y] = new TransportShip(x, y, playerColor, space, parameters);
            return space[Position.X + 1, Position.Y];
        }

        public LocatableObject CreateRadarShip(int price, string[][] parameters)
        {
            resources -= price;
            int x = Position.X + 1, y = Position.Y;
            if (space[Position.X + 1, Position.Y] != null)
                (space[Position.X + 1, Position.Y] as Tactic).Retreat(this);
            space[Position.X + 1, Position.Y] = new RadarShip(x, y, playerColor, space, parameters);
            return space[Position.X + 1, Position.Y];
        }

        public void Captured(PlayerColor color)
        {
            if (capture.Count == 0 || capture.Peek() == color)
                capture.Push(color);
            else
                if (capture.Count > 0)
                    capture.Pop();
            if (capture.Count == 5)
                playerColor = capture.Peek();
        }

        public override void Draw(Graphics canvas, Rectangle canvasSize, Rectangle cursor)
        {
            RectangleF loc = new RectangleF();
            loc.X = canvasSize.Width * ((Position.X - cursor.X) / (float)cursor.Width);
            loc.Y = canvasSize.Height * ((Position.Y - cursor.Y) / (float)cursor.Height);
            loc.Width = canvasSize.Width / (float)cursor.Width;
            loc.Height = canvasSize.Height / (float)cursor.Height;
            canvas.DrawImage(animations.Obj, loc);
        }
    }
}
